Create useful exam mappings derived from exam data

benjamin-shen opened the pull request:
### Summary <!-- Required -->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Itemize bug fixes, new features, and other changes -->
<!-- Feel free to break this into sub-sections, i.e. features, fixes, etc. -->
<!-- Some examples are shown below. -->

This pull request is step 2 in the AP/IB Infra Update #576.

- [x] export college keys/abbreviations from `src/requirements/data`
- [x] added `examRequirementsMapping`
- [x] added `examToCourseMapping`
- [x] added `courseToExamMapping`

<!--- Itemize any relevant remaining TODOs (especially for WIP PRs) here and on Notion -->

<!--- Note dependencies on other PRs if any. -->

### Test Plan <!-- Required -->

<!-- Provide screenshots or point out the additional unit tests -->
![image](https://user-images.githubusercontent.com/34158284/144726963-de0544c4-a68e-41a6-88f6-5896aed2f232.png)
![image](https://user-images.githubusercontent.com/34158284/144726986-ebcef4e9-6f77-4628-b6f0-c42f02e33d6a.png)
![image](https://user-images.githubusercontent.com/34158284/143960937-f533beed-9650-43c5-8bb4-5537a9270d3e.png)


### Notes <!-- Optional -->

<!--- List any important or subtle points, future considerations, or other items of note. -->
Pseudocode for next steps, in the requirements json generation script:  
```
For each requirement in the decorated requirements json:
  course list = requirement[courses]
  Initialize exam id set = {}
  Initialize equivalent course set = {}

  // figure out which exam ids should be added to the course list
  For each course id in course list:
    Associated exam ids = courseToExamMapping[course id]
    If associated exam ids size > 0:
      Add associated exam ids to exam id set
      Add course id to the equivalent course set

  // update course list in requirements json
  Concat list(exam id set) to course list

  // figure out if there are any conditions
  conditions = {}
  For each exam id in exam id set:
    Associated equivalent courses = examToCourseMapping[exam id] intersect equivalent course set
    Initialize valid colleges set = {} // colleges specific to all equivalent courses (i.e. the requirement)
    For each course id in associated equivalent courses:
      Associated valid colleges = examRequirementsMapping[exam id].colleges[course id] // specific to one equivalent course
      Add associated valid colleges to valid colleges set // aggregation of all equivalent courses
    conditions.colleges = list(valid colleges set)
    conditions.excludedMajors = examRequirementsMapping[exam id].excludedMajors
  Add conditions to requirement
```

hahnbeelee commented on src/requirements/requirement-exam-mapping.ts:
```diff
@@ -0,0 +1,122 @@
+  {}
+);
+
+export const examToCourseMapping: Record<number, string[]> = Object.entries(
```
nit - i feel like you can just call it Map instead of Mapping

benjamin-shen replied:
Should I rename the other variables too?

hahnbeelee replied:
eh it's ok u can keep it mapping lol

hahnbeelee commented on src/requirements/requirement-exam-mapping.ts:
```diff
@@ -0,0 +1,122 @@
+  .flat(2);
+
+/**
+ * Mapping from exam id to requirements info (course equivalents and exam conditions), i.e.
```
Nit - change "course equivalents and exam conditions" to "course equivalents and college and major conditions for which the exam fulfills" (or something along those lines)

benjamin-shen replied:
I get your point. I guess we can just call it `courseId` everywhere because there is no ambiguity for what a `courseId` means when associated with an exam.

> So I'm assuming you named this courseId in examData because it's the courseId equivalent for AP/IB? But they all are unique ids for the ap/ib exam it represents?

So yes to the first question, and almost to the second question. It's a unique id for each exam fulfillment (each exam can have more than one fulfillment if there are multiple scores it accepts). This is necessary because we want to use it as a representation of the exam in the requirements data, and different fulfillments (i.e. user scores) can fulfill requirements differently. The unique id for each exam is simply the exam name.

hahnbeelee reviewed (changes requested):
My main issues are with `examRequirementsMapping`. I don't think it's super clear that there's an id for an exam and then corresponding ids for their course equivalents
It took me awhile to understand this:

![image](https://user-images.githubusercontent.com/55263191/144356624-c530f97a-e38c-48f3-a07b-99aacb06fa4b.png)

Like I said I think my main confusion is with the terms used. Specifically the terms default college and the different types of ids.

I also think that it's a little odd that for the `colleges` property it's a mapping from courseId to colleges. The name implies that it should just be an array of colleges, but it's more complex than that.

I also think there's some issues with the logic within `examRequirementsMapping` but I can't 100% confirm these suspicions till you clarify all the definitions for me. Much appreciated! 😁
Hope you enjoy this extremely thorough PR review! 😁

hahnbeelee commented:
Ok as for your pseudocode I think I'm gonna need to take some time to re-read your AP/IB proposal + also re-familiarize myself with the graph algorithm right now. So I'll do that after you address all my comments! lol

benjamin-shen commented on src/requirements/data/exams/ExamCredit.ts:
```diff
@@ -15,6 +15,8 @@ export type ExamFulfillment = {
 export type ExamFulfillments = Record<string, ExamFulfillment[]>;
 export type ExamData = Record<'AP' | 'IB', ExamFulfillments>;
 
+export const DEFAULT_COLLEGES = 'DEFAULT';
```
I think `OTHER_COLLEGES` might make more sense since "default" implies it always takes precedence.

benjamin-shen commented on src/requirements/data/index.ts:
```diff
@@ -399,4 +399,7 @@ const json: RequirementsJson = {
 };
 
+export const colleges = [...Object.keys(json.college)] as const;
+export type College = typeof colleges[number];
```
Not sure what you mean by polymorphic, are you thinking of [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html)?

I think this is just the syntax for turning the list of strings (indexed by number) into a type. I straight up copied [this answer](https://stackoverflow.com/a/54071129).

if `colleges` looks like `['EN', 'AS1', 'AS2']`  
then `College` looks like `'EN' | 'AS1' | 'AS2' | ...`

For example, `college: College` is much better than just saying `college: string`. There are other instances of using `string` as the college type in the codebase but that's not super relevant to this PR.

SamChou19815 replied:
<img width="509" alt="Screen Shot 2021-12-06 at 11 51 05 AM" src="https://user-images.githubusercontent.com/4290500/144887768-28597e75-fc81-49f5-bf99-782717f4dd52.png">

I don't think it achieves what you hoped. I think the only right way to do this is to actually explicitly enumerate all college names

benjamin-shen replied:
😢 you're right. this makes me so sad

![image](https://user-images.githubusercontent.com/34158284/147043299-1c374ad0-4bbf-4451-846e-2b2937da0b4b.png)
Explicitly enumerating the colleges works, so I'll use it for now (hopefully with some future refactoring in `requirements/type.ts`).

benjamin-shen commented on src/requirements/requirement-exam-mapping.ts:
```diff
@@ -0,0 +1,122 @@
+  (mapping, fulfillment) => {
+    const { courseId, courseEquivalents, majorsExcluded } = fulfillment;
+
+    if (!courseEquivalents) {
+      // if no course equivalents, return empty mapping for exam
+      return { ...mapping, [courseId]: {} };
+    }
```
Hmm I think you're right. If an exam has no course equivalents, we would never look them up when adding data to the requirements json. Thank you for the catch!

benjamin-shen commented on src/requirements/requirement-exam-mapping.ts:
```diff
@@ -0,0 +1,122 @@
+import examData, {
```
I could! Although I think it might be better to keep complicated logic out of the `data/` directory. i.e. if we want to update the AP/IB exam data, we'd go to `data/ExamCredit.ts`; if we want to update the mappings logic, we'd go to this file.

benjamin-shen commented on src/requirements/requirement-exam-mapping.ts:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    // for each id, assign a list of colleges for which the exam can fulfill requirements
+    const nonDefaultColleges = Object.keys(courseEquivalents);
```
I see what you mean. The variable's value is correct but I think it can be renamed `definedColleges` or something. Whether or not the actual string `'DEFAULT'` is in there doesn't matter, since it's just being used as a mask over `colleges`.

benjamin-shen commented:
@hahnbeelee that's a great diagram 😀 

> Like I said I think my main confusion is with the terms used. Specifically the terms default college and the different types of ids.

I changed `DEFAULT_COLLEGE` to `OTHER_COLLEGES`. I got rid of `examId` entirely because it's the same as `courseId`.

> I also think that it's a little odd that for the colleges property it's a mapping from courseId to colleges. The name implies that it should just be an array of colleges, but it's more complex than that.

I changed `colleges` to `collegeConditions` and hopefully named the types a little bit better.

> I also think there's some issues with the logic within examRequirementsMapping but I can't 100% confirm these suspicions till you clarify all the definitions for me. Much appreciated! 😁

Totally get that, lmk if you have any more questions! I also have suspicions related to the `majorsExcluded` but I think that can definitely be addressed in a later PR. It's such a smol edge case, I'm mainly concerned about getting the college criteria to work properly.

SamChou19815 commented on src/requirements/requirement-exam-mapping.ts:
```diff
@@ -0,0 +1,142 @@
+  };
+}, {});
+
+export const courseToExamMapping: Record<number, string[]> = Object.entries(
+  examToCourseMapping
+).reduce((mapping: Record<number, string[]>, [id, courses]) => {
+  courses
+    .map(c => parseInt(c, 10))
+    .forEach(course => {
+      if (!mapping[course]) mapping[course] = [];
+      mapping[course] = [...mapping[course], id];
+    });
+  return mapping;
+}, {});
```
Maybe using `Object.fromEntries` can be cleaner.

hahnbeelee commented on src/requirements/data/exams/ExamCredit.ts:
```diff
@@ -15,7 +15,8 @@ export type ExamFulfillment = {
 export type ExamData = Record<'AP' | 'IB', ExamFulfillments>;
 
-export const DEFAULT_COLLEGES = 'DEFAULT';
+// If the user's college is not in the keys of courseEquivalents, their college is generalized to OTHER_COLLEGES
```
oh lol this makes sense

hahnbeelee reviewed (approved):
LGTM! It's much clearer what's going on now. Thank you for renaming the variables!

hahnbeelee reviewed (approved):
very clean changes lgtm!
![image](https://user-images.githubusercontent.com/55263191/147368432-f21fc95f-b36f-4704-8505-4df64ce14211.png)
